lib/pbench/server/database/models/users.py

```diff
@@ -57,107 +55,44 @@ def __init__(self, user: "User", cause: str):
-        self.key = key
-
-    def __str__(self) -> str:
-        return f"User profile key {self.key!r} is not supported"
-
-
-class UserProfileBadStructure(UserError):
-    """Invalid Profile structure encountered.
-
-    A call to update found a level in the JSON document where
-    the caller's key expected a nested JSON object but the type at that level
-    is something else. For example, when `profile.user.email` finds that
-    `profile.user` is a string, it's impossible to look up the `email` field.
-
-    The error text will identify the key path and the expected key element that
-    is missing.
-    """
-
-    def __init__(self, key: str):
-        self.key = key
-
-    def __str__(self) -> str:
-        return f"Key value for {self.key!r} in profile is not a JSON object"
+        return f"Missing required key in {self.user.get_json()}: {self.cause}"
 
 
 class User(Database.Base):
     """User Model for storing user related details."""
 
     __tablename__ = "users"
 
-    # User profile keys
-    #
-    # SERVER is an internally maintained key namespace for additional metadata
-    # relating to the server's management of user. The information here is
-    # accessible to callers, but can't be changed.
-    #
-    # {"server.role": ["ADMIN"]}
-    #
-    # USER provides an open namespace allowing a user which is authenticated
-    # to define arbitrary metadata accessible only to that authenticated user.
-    # Each user can have its own unique value for this key, for example
-    # "user.email"/"user.first_name".
-    # Unauthenticated users can neither set nor read the user profile
-    # in the USER namespace.
-    # Note: USER namespace will be populated by the Server when the user access
-    # the Pbench Server resources for the very first time.
-    #
-    # {"user.email": user@email.com}
-
-    SERVER = "server"
-    USER = "user"
-
-    # Keys that are updatable by the user
-    USER_UPDATEABLE_KEYS = [USER]
-
-    # Keys that authenticated user can read
-    PROFILE_KEYS = [USER, SERVER]
-
-    _valid_key_charset = re.compile(r"[a-z0-9_.-]+")
-
     id = Column(Integer, primary_key=True, autoincrement=True)
     oidc_id = Column(String(255), unique=True, nullable=False)
     username = Column(String(255), unique=True, nullable=True)
-    profile = Column(JSON, unique=False, nullable=True)
     dataset_metadata = relationship(
         "Metadata", back_populates="user", cascade="all, delete-orphan"
     )
+    _roles = Column(String(255), unique=False, nullable=True)
+
+    @property
+    def roles(self):
+        if self._roles:
+            return [x for x in self._roles.split(";")]
+        else:
+            return []
+
+    @roles.setter
+    def roles(self, value):
+        if isinstance(value, list):
+            for v in value:
+                self._roles += f";{v}"
+        elif isinstance(value, str):
+            self._roles = value if not self._roles else f"{self._roles};{value}"
```

Comment:
> Don't we want to set this to the values decoded from the token?

Given my understanding of the current direction in which we seem to be heading, I think we do.  🥴

Reply:
Yeah I do add the roles from the decoded oidc access token, Its done under auth.py line 215

> So each `User.roles = "a"` _appends_ another role? That sounds ... odd. It seems there's no way to _clear_ it or _update_ it.

Yeah, you are right the current approach is just _adding_ a new role and there is no way to clear a specific role if we want to. I started down this approach and didn't realize there is no "removing the role" functionality but if all the outstanding comments are resolved and we decided to have the `remove_role` functionality I will update the `user.update` method.

Reply:
Your `Auth` code doesn't currently *update* the cache if the oidc_id is found, and for creating a new model proxy I don't think it matters. But I suspect we *should* be updating username and roles if they differ from what's already there; and both for that purpose and the existing use, setting `roles=...` should *replace* the current current list, not try to edit it. We have no use case where we want to add or remove individual roles ... we simply want the cached roles to reflect the Keycloak roles from the token.

Reply:
Yep, I agree I should just replace the roles instead of editing the existing ones. and also I should add a logic that if the user is present then check if any updates need to be performed.

Reply:
> I suspect we should be updating username and roles if they differ from what's already there

I affirm that:  the user doesn't necessarily control the roles (and the token is supposed to be secure), so, when the user presents the token, it enables the ID provider to tell the Pbench Server what roles the user has been assigned.  This is important, because, between the last time we saw the user and this time, the user may have been "demoted" (i.e., they may have lost a role which they had previously), and we'll know this when we extract the roles from the token.

So, on each presentation of a token, we need to update the username and roles, because either might have changed.  (Presumably, those changes will be quite rare, but we need to be prepared for them.)

As to whether it is worth trying to determine whether an update is necessary before we go ahead and do it is a different question.  I would kind of hope that that sort of optimization is already baked into the ORM -- if we try to set an object to a value that it already has and then try to commit that to the database, I might hope that the engine would just wink and nod and not actually perform the transaction (assuming that there are no side effects).

> We have no use case where we want to add or remove individual roles ... we simply want the cached roles to reflect the Keycloak roles from the token.

Yep...I believe that that is the model that we're embracing.  (I'm not confident that it's going to work all that well, but it's a good place to start.)